src/optimization.py

```diff
@@ -0,0 +1,363 @@
+        return None
+
+
+class LeastSquares(Optimization):
+
+    def __init__(self,
+                 covariance: Optional[Covariance] = None,
+                 *arg, **kwarg):
+        super().__init__(*arg, **kwarg)
+        self.covariance = covariance
+
+    def set_objective(self, optimization_data: OptimizationData) -> None:
+        X, y = optimization_data.view(self.constraints.selection, mode='log')
+
+        # 0.5 * w * P * w' - q * w' + constant
+        P = 2 * (X.T @ X)
+        q = to_numpy(-2 * X.T @ y).reshape((-1,))
+        constant = to_numpy(y.T @ y).item()
+
+        l2_penalty = self.params.get('l2_penalty')
+        if l2_penalty is not None and l2_penalty != 0:
+            P += 2 * l2_penalty * np.eye(X.shape[1])
+
+        self.objective = Objective(P=P, q=q, constant=constant)
+        return None
+
+    def solve(self) -> bool:
+        return super().solve()
+
+
+class WeightedLeastSquares(Optimization):
+
+    def set_objective(self, optimization_data: OptimizationData) -> None:
+        X, y = optimization_data.view(self.constraints.selection, mode='log')
+
+        tau = self.params['tau']
+        lambda_val = np.exp(-np.log(2) / tau)
+        i = np.arange(X.shape[0])
+        wt_tmp = lambda_val ** i
+        wt = np.flip(wt_tmp / np.sum(wt_tmp) * len(wt_tmp))
+        W = np.diag(wt)
+
+        P = 2 * ((X.T).to_numpy() @ W @ X)
+        q = -2 * (X.T).to_numpy() @ W @ y
+        constant = (y.T).to_numpy() @ W @ y
+
+        self.objective = Objective(P=P, q=q, constant=constant)
+        return None
+
+    def solve(self) -> bool:
+        return super().solve()
+
+
+class QEQW(Optimization):
+
+    def __init__(self, *arg, **kwarg):
+        covariance = Covariance(method='duv')
+        super().__init__(covariance=covariance, *arg, **kwarg)
+
+    def set_objective(self, optimization_data: OptimizationData) -> None:
+        covmat = self.covariance.estimate(X=optimization_data['X']) * 2
+        mu = np.zeros(optimization_data['X'].shape[1])
+        self.objective = Objective(P=covmat, q=mu)
+        return None
+
+    def solve(self) -> bool:
+        return super().solve()
+
+
+class LAD(Optimization):
+    # Least Absolute Deviation (same as mean absolute deviation, MAD)
+
+    def __init__(self, *arg, **kwarg):
+        super().__init__(*arg, **kwarg)
+        self.params['use_level'] = self.params.get('use_level', True)
+        self.params['use_log'] = self.params.get('use_log', True)
+
+    def set_objective(self, optimization_data: OptimizationData) -> None:
+        # X = np.log(1 + training_data['X_train'])
+        # y = np.log(1 + training_data['X_bm'])
+        # X = training_data['X_train']
+        # y = training_data['X_bm'].squeeze()
+        # if self.params.get('use_level'):
+        #     X = (1 + X).cumprod()
+        #     y = (1 + y).cumprod()
+        #     if self.params.get('use_log'):
+        #         X = np.log(X)
+        #         y = np.log(y)
```

Comment:
If those lines of are useful, you could add some comments to explain why, otherwise you can delete them